jam/server/api/v2/base.py

```diff
@@ -0,0 +1,296 @@
+    }
+
+    def get_current_user(self):
+        try:
+            return User(
+                self.request.headers.get('Authorization') or
+                self.get_query_argument('token', default=None)
+            )
+        except jwt.ExpiredSignatureError:
+            return User(None)
+
+    def serialize(self, inst):
+        return self._serializer.serialize(self.request, inst, *self._view.parents)
+
+    def initialize(self, view, serializer):
+        self._view = None
+        self._view_class = view
+        self._serializer = serializer
+
+    def prepare(self):
+        super().prepare()
+        if self.request.method == 'OPTIONS':
+            return  # Dont do anything for OPTIONS requests
+
+        # if type(self.json) not in self.TYPE_DATA_SET[self.request.method]:
+        #     raise Exception()
+
+        loaded = []
+        try:
+            for view in self._view_class.lineage():
+                key = view.name + '_id'
+                if self.path_kwargs[key] is None:
+                    break
+                loaded.append(view.load(self.path_kwargs[key], *loaded))
+        except exceptions.NotFound as e:
+            err = e
+        else:
+            err = None
+
+        self._view = self._view_class(*loaded)
+
+        if 'relationship' in self.path_kwargs:
+            relationship = self._serializer.relations[self.path_kwargs['relationship']]
+            self._view = relationship.view(*loaded)
+            self._serializer = relationship.serializer()
+
+        permissions = Permissions.get_permissions(self.current_user, *loaded)
+        required_permissions = self._view.get_permissions(self.request)
+
+        # For use later on
+        self.current_user.permissions = permissions
+
+        if (required_permissions & permissions) != required_permissions:
+            if self.current_user.uid is None:
+                raise exceptions.Unauthorized()
+            raise exceptions.Forbidden(required_permissions)
+
+        if err:
+            raise err
+
+        if self.request.method in ('GET', 'DELETE'):
+            return  # GET and DELETE bodies are ignored
+
+        if not isinstance(self.json, dict):
+            raise exceptions.MalformedData()
+
+        data = self.json.get('data', {})
+        if data.get('type') != self._view.plural:
+            raise exceptions.IncorrectParameter('data.type', self._view.type, data.get('type', 'null'))
+
+    # Create
+    def post(self, **args):
+        if self._view.resource:
+            raise tornado.web.HTTPError(http.client.METHOD_NOT_ALLOWED)
+
+        if isinstance(self.json['data'], list):
+            return self.post_bulk()
+
+        new = self._view.create(self.json['data'].get('id'), self.json['data']['attributes'], self.current_user)
+        self.write({'data': self.serialize(new)})
+        self.set_status(http.client.CREATED)
+
+    def post_bulk(self):
+        if 'bulk' not in self.extensions:
+            raise Exception()
```

Comment:
Maybe a custom exception that gets passed back to the user telling them to do it correctly next time?